### PR TITLE
Backend: Searching & Filtering

### LINKED ISSUE #48: [Searching & Filtering] Backend
**Description**

- API endpoint for basic searching and filtering of all posts

### PR DESCRIPTION
**Pull request by minhduc29**
# Backend: Searching & Filtering

## Description

- Strengthen interface field types and add missing edge cases for create post functionality
- Fix some minor bugs with overwritten files/outdated PR merge
- Update schema, model, and interface for post
- Add a backend API endpoint for search/filter functionality

Fixes #48 

## How Has This Been Tested

- The database was first populated with a variety of data
- Each type of filter (location, price, rules, etc.) was then tested individually with different combinations and with other types of filters

## Additional context
image
image

## Notes

- The interface for the search params/request body is a bit different from other functionalities
- The filter is minimal compared to the full post data (some fields are grouped together or are not filtered)
- There is a difference between undefined and false for each rule field
- This backend functionality only supports basic searching/filtering, the sorting/suggestion algorithm is implemented in the frontend among the filtered posts

### CODE REVIEW

File: server/src/db/db.ts [hunk 1]
```diff
@@ -3,10 +3,6 @@ import mongoose from 'mongoose';
 const db = async () => {
   try {
     mongoose.set('strictQuery', false);
-
```
Comment (linhle15-ll) [inline server/src/db/db.ts:6]: why remove this?

Comment (minhduc29) [reply server/src/db/db.ts:6]: Your original code from PR #24

**Review by linhle15-ll — COMMENTED** [review]
><
That's pretty much work!
I have some questions in the comments for more clarification.
Can you try create some posts with different information and show how is the basic search and filter work (eg: outcome when search by state and filter by the filter properites). The pics in the additional context showed error: missing location only.

File: server/src/utils/validators.ts [hunk 2]
```diff
@@ -1,6 +1,8 @@
 export const validateMedia = (data: string[]) => data.length >= 5;
 
 export const validateTime = (data: string) => {
+  if (!data) return true; // Time is optional
```
[the remark this answers is not in the document]
Comment (minhduc29) [reply server/src/utils/validators.ts:4]: Without this part, the code in the controller would have been longer (original code)

File: server/src/types/post.types.ts [hunk 3]
```diff
@@ -83,6 +82,6 @@ export interface Post extends Base, Timestamps {
 
 export interface PostRequestBody extends Post {
   address: string;
-  lat: number;
-  long: number;
+  minPrice: number;
```
[the remark this answers is not in the document]
Comment (minhduc29) [reply server/src/types/post.types.ts:85]: My thought process is that when users try to filter for posts within a price range, there will be a min price and a max price. That's the only use case of these fields. For other cases (create/update), these fields won't be used. The most intuitive way would be to have another interface for each particular endpoint but that would be unnecessary at the current scope.

**Comment by minhduc29** [comment]
> That's pretty much work!
> I have some questions in the comments for more clarification.
> Can you try create some posts with different information and show how is the basic search and filter work (eg: outcome when search by state and filter by the filter properites). The pics in the additional context showed error: missing location only.

The additional context is only a few examples of showing how you can test, what the output looks like, and edge cases, if any. I've stated that I've tested all fields individually and in combinations with other fields as well. It is unnecessary to include 20+ photos in the additional context. If you would like to test it on your own, you can follow the description in "How Has This Been Tested". If you want me to write the thorough unit tests, that would also be a different scenario.

**Comment by minhduc29** [comment]
More detailed testing with location filtering:
- 3 posts with location data:
image
image
image
- Request with zip "22038" returns 2 objects:
image
- Request with state "CA" (and city "Hehe") returns 1 object:
image
- Request with lat 167.5 and long 175.9 returns 2 objects:
image
- Requests with other information that doesn't match any posts don't return any objects:
image

**Review by linhle15-ll — APPROVED** [review]
